app-libs/sgx-runtime/pallets/guess-the-number/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+        )]
+		pub fn guess(origin: OriginFor<T>, guess: GuessType) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(
+				Self::guess_attempts(&sender) < T::MaxAttempts::get(),
+				Error::<T>::TooManyAttempts
+			);
+			<GuessAttempts<T>>::mutate(&sender, |a| *a += 1u8);
+			let lucky_number = <LuckyNumber<T>>::get().ok_or(Error::<T>::NoDrawYet)?;
+			let distance = GuessType::abs_diff(lucky_number, guess);
+			let last_winning_distance = Self::winning_distance().unwrap_or(crate::GuessType::MAX);
+			if distance <= last_winning_distance {
+				let mut winners =
+					if distance == last_winning_distance { <Winners<T>>::get() } else { vec![] };
+				<WinningDistance<T>>::put(distance);
+				ensure!(winners.len() < T::MaxWinners::get() as usize, Error::<T>::TooManyWinners);
+				if !winners.contains(&sender) {
+					winners.push(sender);
+					Winners::<T>::put(winners);
+				}
+			}
+			Ok(().into())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T>
+where
+	sp_core::H256: From<<T as frame_system::Config>::Hash>,
+{
+	pub fn get_pot_account() -> T::AccountId {
+		let pot_identifier = <T as Config>::PalletId::get();
+		let pot_id_hash: H256 = T::Hashing::hash_of(&pot_identifier.0.as_slice()).into();
+		T::AccountId::decode(&mut pot_id_hash.as_bytes())
+			.expect("32 bytes can always construct an AccountId32")
+	}
+
+	pub fn payout_winners() {
+		let pot = Self::get_pot_account();
+		let winners = Self::winners();
+		let winnings = min(Self::winnings(), T::Currency::free_balance(&pot));
+		let winnings_per_winner = winnings
+			.checked_div(&BalanceOf::<T>::saturated_from(winners.len() as u32))
+			.unwrap_or_default();
+
+		for winner in winners {
+			if T::Currency::transfer(
+				&pot,
+				&winner,
+				winnings_per_winner,
+				ExistenceRequirement::AllowDeath,
+			)
+			.is_err()
+			{
+				warn!("error transferring rewards")
+			};
+		}
+	}
+
+	fn progress_round() -> DispatchResult {
+		let current_round_index = <CurrentRoundIndex<T>>::get();
+		let last_round_timestamp = Self::next_round_timestamp();
+
+		Self::end_round()?;
+
+		let next_round_index = current_round_index.saturating_add(1);
+		<CurrentRoundIndex<T>>::put(next_round_index);
+		info!("new round with index {}", next_round_index);
+
+		let next = last_round_timestamp.saturating_add(T::RoundDuration::get());
+		<NextRoundTimestamp<T>>::put(next);
+
+		Self::start_round()
+	}
+
+	fn end_round() -> DispatchResult {
+		let current_round_index = <CurrentRoundIndex<T>>::get();
+		info!("ending round {}", current_round_index);
+		Self::payout_winners();
+		<LastWinners<T>>::put(Self::winners());
+		<Winners<T>>::kill();
+		<LastWinningDistance<T>>::put(Self::winning_distance().unwrap_or(GuessType::MAX));
+		<WinningDistance<T>>::kill();
+		<LastLuckyNumber<T>>::put(Self::lucky_number().unwrap_or(0));
+		let _ = <GuessAttempts<T>>::clear(T::MaxAttempts::get().into(), None);
```

Comment:
I am not sure if I remember the function correctly, does this really clear all the quesses?

Reply:
hmm, that limit argument is actually wrong. good spot.